File: scripts/indec/ipc/update_ipc.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from 'url';
import XLSX from 'xlsx';

const BASE_URL = "https://www.indec.gob.ar/ftp/cuadros/economia";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const BACKEND_ROOT = path.resolve(__dirname, '../../..');
const DATA_DIR = path.join(BACKEND_ROOT, 'data', 'ipc');

function armarNombreArchivo(anio, mes) {
  const mm = String(mes).padStart(2, "0");
  const aa = String(anio).slice(-2);

  return `sh_ipc_${mm}_${aa}.xls`;
}

function armarUrl(anio, mes) {
  return `${BASE_URL}/${armarNombreArchivo(anio, mes)}`;
}

async function descargarArchivo(url) {
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`No se pudo descargar. Estado HTTP: ${response.status}`);
  }

  const arrayBuffer = await response.arrayBuffer();
  return Buffer.from(arrayBuffer);
}

async function actualizarIpc(anio, mes) {
  const archivo = armarNombreArchivo(anio, mes);
  const url = armarUrl(anio, mes);

  console.log("Descargando IPC desde INDEC...");
  console.log(`URL: ${url}`);

  const buffer = await descargarArchivo(url);

  console.log("");
  console.log("Archivo descargado correctamente y parseado en memoria:");
  console.log(archivo);

  // Parsear el archivo descargado y exportar CSV largo
  try {
    await parsearIpcDesdeBuffer(buffer, archivo);
  } catch (err) {
    console.error('Error en parseo posterior a la descarga:', err.message);
  }
}

const [, , anioArg, mesArg] = process.argv;

if (!anioArg || !mesArg) {
  console.log("Uso:");
  console.log("node scripts/indec/ipc/update_ipc.js 2026 4");
  process.exit(1);
}

const anio = Number(anioArg);
const mes = Number(mesArg);

if (!anio || !mes || mes < 1 || mes > 12) {
  console.error("Año o mes inválido.");
  process.exit(1);
}

actualizarIpc(anio, mes).catch((error) => {
  console.error("");
  console.error("Error actualizando IPC:");
  console.error(error.message);
  process.exit(1);
});

// -------------------- parser (adaptado de parse_ipc.js) --------------------

const OUTPUT_DIR = path.join(BACKEND_ROOT, 'src', 'data', 'ipc');
const OUTPUT_FILE = path.join(OUTPUT_DIR, 'ipc_aperturas_largo.csv');
const OUTPUT_FILE_INTERANUAL = path.join(OUTPUT_DIR, 'ipc_interanual_largo.csv');

const CONCEPTOS_VALIDOS = [
  "Nivel general",
  "Alimentos y bebidas no alcohólicas",
  "Bebidas alcohólicas y tabaco",
  "Prendas de vestir y calzado",
  "Vivienda, agua, electricidad, gas y otros combustibles",
  "Equipamiento y mantenimiento del hogar",
  "Salud",
  "Transporte",
  "Comunicación",
  "Recreación y cultura",
  "Educación",
  "Restaurantes y hoteles",
  "Bienes y servicios varios",
  "Estacional",
  "Núcleo",
  "Regulados",
  "Bienes",
  "Servicios",
];

const REGEX_FECHA = /^\d{2}\/\d{2}$/;

function normalizarTexto(valor) {
  return String(valor || "").replace(/\s+/g, " ").trim();
}

function limpiarConcepto(valor) {
  return normalizarTexto(valor).replace(/\*+$/g, "").trim();
}

function convertirFecha(valor) {
  if (valor instanceof Date) {
    const mm = String(valor.getMonth() + 1).padStart(2, "0");
    const aa = String(valor.getFullYear());
    return `${aa}-${mm}`;
  }

  if (typeof valor === "number") {
    // Excel serial date to JS Date
    // ignore small serials (0, empty) that map to 1899/1900
    if (valor < 30000) return null;
    const d = new Date(Math.round((valor - 25569) * 86400 * 1000));
    const mm = String(d.getMonth() + 1).padStart(2, "0");
    const aa = String(d.getFullYear());
    return `${aa}-${mm}`;
  }
  if (typeof valor === 'string') {
    const texto = normalizarTexto(valor);

    // Try parse as mm/yy or m/yy
    const m = texto.match(/^(\d{1,2})\/(\d{2})$/);
    if (m) {
      const mes = String(m[1]).padStart(2, "0");
      const anio = `20${m[2]}`;
      return `${anio}-${mes}`;
    }

    // Try JS Date parse on strings like 'Sun Jan 01 2017 ...'
    const parsed = Date.parse(texto);
    if (!Number.isNaN(parsed)) {
      const d = new Date(parsed);
      const mm = String(d.getMonth() + 1).padStart(2, "0");
      const aa = String(d.getFullYear());
      return `${aa}-${mm}`;
    }

    return null;
  }

  return null;
}

function convertirNumero(valor) {
  if (valor === null || valor === undefined || valor === "") return null;
  if (typeof valor === "number") return valor;

  const texto = String(valor).replace(/\./g, "").replace(",", ".").trim();
  const numero = Number(texto);

  return Number.isNaN(numero) ? null : numero;
}

function detectarTipoVariacion(sheetData) {
  const texto = sheetData.slice(0, 10).flat().map(normalizarTexto).join(" ").toLowerCase();

  if (texto.includes("interanual")) return "interanual";
  if (texto.includes("mensual")) return "mensual";
  if (texto.includes("acumulad")) return "acumulada";

  return "sin_identificar";
}

function esRegion(texto) {
  const t = normalizarTexto(texto);

  return (
    t === "Total nacional" ||
    t.startsWith("Región") ||
    ["GBA","Pampeana","Noreste","Noroeste","Cuyo","Patagonia"].includes(t)
  );
}

function detectarBloque(texto) {
  const t = normalizarTexto(texto).toLowerCase();

  if (t.includes("nivel general") || t.includes("divisiones")) {
    return "Nivel general";
  }

  if (t.includes("categor")) {
    return "Categorías";
  }

  if (t.includes("bienes y servicios")) {
    return "Bienes y servicios";
  }

  return null;
}

function encontrarColumnasFecha(fila) {
  const columnas = [];

  fila.forEach((celda, index) => {
    const fecha = convertirFecha(celda);

    if (fecha) {
      columnas.push({ index, fecha });
    }
  });

  return columnas;
}

function encontrarConceptoEnFila(fila) {
  for (const celda of fila) {
    const texto = limpiarConcepto(celda);

    if (CONCEPTOS_VALIDOS.includes(texto)) {
      return texto;
    }
  }

  return null;
}

function csvEscape(valor) {
  const texto = String(valor ?? "");

  if (texto.includes(",") || texto.includes('"') || texto.includes("\n")) {
    return `"${texto.replace(/"/g, '""')}"`;
  }

  return texto;
}

function guardarCsv(rows) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  const header = [
    "fecha",
    "region",
    "tipo_variacion",
    "bloque",
    "concepto",
    "valor",
    "archivo_origen",
    "fecha_descarga",
  ];

  const lines = rows.map((row) =>
    [
      row.fecha,
      row.region,
      row.tipo_variacion,
      row.bloque,
      row.concepto,
      row.valor,
      row.archivo_origen,
      row.fecha_descarga,
    ]
      .map(csvEscape)
      .join(",")
  );

  fs.writeFileSync(OUTPUT_FILE, [header.join(","), ...lines].join("\n"), "utf8");
}

async function parsearIpcDesdeBuffer(buffer, archivoOrigen) {
  if (!buffer || !buffer.length) {
    console.error("No existe contenido descargado para parsear.");
    return;
  }

  const FECHA_DESCARGA = new Date().toISOString().slice(0, 10);

  const workbook = XLSX.read(buffer, { type: 'buffer' });

  // helper to parse one sheet and write to given output file
  function parseSheetToCsv(sheetName, outFile) {
    const sheet = workbook.Sheets[sheetName];
    const sheetData = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: "" });

    const tipoVariacion = detectarTipoVariacion(sheetData);

    let regionActual = "";
    let bloqueActual = "";
    let columnasFechaActuales = [];

    const resultados = [];

    for (let rowIndex = 0; rowIndex < sheetData.length; rowIndex++) {
      const fila = sheetData[rowIndex];

      const columnasFecha = encontrarColumnasFecha(fila);

      if (columnasFecha.length >= 3) {
        columnasFechaActuales = columnasFecha;
      }

      const textosFila = fila.map(normalizarTexto).filter(Boolean);

      for (const texto of textosFila) {
        if (esRegion(texto)) {
          regionActual = texto;
          bloqueActual = "";
        }

        const bloqueDetectado = detectarBloque(texto);

        if (bloqueDetectado) {
          bloqueActual = bloqueDetectado;
        }
      }

      const concepto = encontrarConceptoEnFila(fila);

      if (!concepto) continue;
      if (!regionActual) continue;
      if (!bloqueActual) continue;
      if (columnasFechaActuales.length === 0) continue;

      for (const col of columnasFechaActuales) {
        const valor = convertirNumero(fila[col.index]);

        if (valor === null) continue;

        resultados.push({
          fecha: col.fecha,
          region: regionActual,
          tipo_variacion: tipoVariacion,
          bloque: bloqueActual,
          concepto,
          valor,
          archivo_origen: archivoOrigen,
          fecha_descarga: FECHA_DESCARGA,
        });
      }
    }

    // write to outFile
    fs.mkdirSync(path.dirname(outFile), { recursive: true });

    const header = [
      "fecha",
      "region",
      "tipo_variacion",
      "bloque",
      "concepto",
      "valor",
      "archivo_origen",
      "fecha_descarga",
    ];

    const lines = resultados.map((row) =>
      [
        row.fecha,
        row.region,
        row.tipo_variacion,
        row.bloque,
        row.concepto,
        row.valor,
        row.archivo_origen,
        row.fecha_descarga,
      ]
        .map(csvEscape)
        .join(",")
    );

    fs.writeFileSync(outFile, [header.join(","), ...lines].join("\n"), "utf8");

    console.log("Parse IPC terminado.");
    console.log(`Hoja leída: ${sheetName}`);
    console.log(`Tipo de variación detectada: ${tipoVariacion}`);
    console.log(`Filas generadas: ${resultados.length}`);
    console.log(`Archivo generado: ${outFile}`);
  }

  // parse first sheet -> aperturas
  if (workbook.SheetNames.length > 0) {
    parseSheetToCsv(workbook.SheetNames[0], OUTPUT_FILE);
  }

  // if second sheet exists, parse it as interanual
  if (workbook.SheetNames.length > 1) {
    const name1 = workbook.SheetNames[1];
    // prefer sheet with 'interanual' in name, otherwise still parse second sheet
    if (name1.toLowerCase().includes('interanual') || true) {
      parseSheetToCsv(name1, OUTPUT_FILE_INTERANUAL);
    }
  }
}

File: scripts/indec/ipc/update_ipc.py

import argparse
import csv
import re
import subprocess
from datetime import datetime
from io import BytesIO
from pathlib import Path

import pandas as pd
import requests

BASE_URL = "https://www.indec.gob.ar/ftp/cuadros/economia"

BACKEND_ROOT = Path(__file__).resolve().parents[3]
DATA_DIR = BACKEND_ROOT / "data" / "ipc"
OUTPUT_DIR = BACKEND_ROOT / "src" / "data" / "ipc"
OUTPUT_FILE = OUTPUT_DIR / "ipc_aperturas_largo.csv"
OUTPUT_FILE_INTERANUAL = OUTPUT_DIR / "ipc_interanual_largo.csv"

CONCEPTOS_VALIDOS = {
    "Nivel general",
    "Alimentos y bebidas no alcohólicas",
    "Bebidas alcohólicas y tabaco",
    "Prendas de vestir y calzado",
    "Vivienda, agua, electricidad, gas y otros combustibles",
    "Equipamiento y mantenimiento del hogar",
    "Salud",
    "Transporte",
    "Comunicación",
    "Recreación y cultura",
    "Educación",
    "Restaurantes y hoteles",
    "Bienes y servicios varios",
    "Estacional",
    "Núcleo",
    "Regulados",
    "Bienes",
    "Servicios",
}


def armar_nombres_posibles(anio, mes):
    yy = str(anio)[-2:]

    return [
        f"sh_ipc_{mes:02d}_{yy}.xls",
        f"sh_ipc_{mes:02d}_{yy}.xlsx",
        f"ipc_{mes:02d}_{yy}.xls",
        f"ipc_{mes:02d}_{yy}.xlsx",
        f"sh_ipc_{mes}_{yy}.xls",
        f"sh_ipc_{mes}_{yy}.xlsx",
    ]


def es_excel_valido(content: bytes) -> bool:
    if not content or len(content) < 1000:
        return False

    return (
        content.startswith(b"\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1")
        or content.startswith(b"PK\x03\x04")
    )


def descargar_archivo(url: str) -> bytes:
    print(f"Intentando descargar: {url}")

    try:
        response = requests.get(
            url,
            timeout=90,
            proxies={"https": "", "http": ""},
            headers={"User-Agent": "Mozilla/5.0"},
        )

        print("Status requests:", response.status_code)
        response.raise_for_status()

        if not es_excel_valido(response.content):
            raise RuntimeError("La respuesta no parece ser un Excel válido.")

        return response.content

    except Exception as err:
        print("Falló requests:")
        print(err)

    try:
        print("Intentando con curl...")
        result = subprocess.run(
            [
                "curl.exe",
                "--silent",
                "--show-error",
                "--location",
                "--fail",
                url,
            ],
            capture_output=True,
            check=True,
        )

        if not es_excel_valido(result.stdout):
            raise RuntimeError("La descarga por curl no parece ser un Excel válido.")

        return result.stdout

    except Exception as err:
        print("Falló curl:")
        print(err)
        raise RuntimeError(f"No se pudo descargar el archivo desde: {url}") from err


def buscar_y_descargar_ipc(anio: int, mes: int) -> tuple[str, bytes]:
    print("Descargando IPC desde INDEC...")

    errores = []

    for archivo in armar_nombres_posibles(anio, mes):
        url = f"{BASE_URL}/{archivo}"
        print(f"\nProbando URL: {url}")

        try:
            content = descargar_archivo(url)
            print(f"[OK] Archivo encontrado: {archivo}")
            return archivo, content
        except Exception as err:
            errores.append(f"{archivo}: {err}")
            print(f"[WARN] No se pudo descargar {archivo}")

    print("\nNo se encontró archivo IPC válido. Intentos realizados:")
    for err in errores:
        print("-", err)

    raise RuntimeError("No se encontró ningún archivo IPC válido para ese mes.")


def normalizar_texto(valor) -> str:
    return re.sub(r"\s+", " ", str(valor or "")).strip()


def limpiar_concepto(valor) -> str:
    return re.sub(r"\*+$", "", normalizar_texto(valor)).strip()


def convertir_fecha(valor) -> str | None:
    if valor is None or (isinstance(valor, float) and pd.isna(valor)):
        return None

    if isinstance(valor, datetime):
        return valor.strftime("%Y-%m")

    if isinstance(valor, pd.Timestamp):
        if pd.isna(valor):
            return None
        return valor.strftime("%Y-%m")

    if isinstance(valor, (int, float)):
        if valor < 30000:
            return None
        try:
            d = pd.to_datetime(valor, unit="D", origin="1899-12-30")
            return d.strftime("%Y-%m")
        except Exception:
            return None

    if isinstance(valor, str):
        texto = normalizar_texto(valor)
        m = re.match(r"^(\d{1,2})/(\d{2})$", texto)
        if m:
            mes = int(m.group(1))
            anio = int(m.group(2))
            if 1 <= mes <= 12:
                return f"20{anio:02d}-{mes:02d}"

        try:
            parsed = pd.to_datetime(texto)
            if pd.isna(parsed):
                return None
            return parsed.strftime("%Y-%m")
        except Exception:
            return None

    return None


def convertir_numero(valor) -> float | None:
    if valor is None or (isinstance(valor, float) and pd.isna(valor)):
        return None

    if isinstance(valor, (int, float)):
        return float(valor)

    texto = str(valor).replace(".", "").replace(",", ".").strip()

    if not texto:
        return None

    try:
        return float(texto)
    except ValueError:
        return None


def detectar_tipo_variacion(sheet_data: list[list]) -> str:
    top = " ".join(
        normalizar_texto(c)
        for fila in sheet_data[:10]
        for c in fila
    ).lower()

    if "interanual" in top:
        return "interanual"

    if "mensual" in top:
        return "mensual"

    if "acumulad" in top:
        return "acumulada"

    return "sin_identificar"


def es_region(texto: str) -> bool:
    t = normalizar_texto(texto)

    return t == "Total nacional" or t.startswith("Región") or t in {
        "GBA",
        "Pampeana",
        "Noreste",
        "Noroeste",
        "Cuyo",
        "Patagonia",
    }


def detectar_bloque(texto: str) -> str | None:
    t = normalizar_texto(texto).lower()

    if "nivel general" in t or "divisiones" in t:
        return "Nivel general"

    if "categor" in t:
        return "Categorías"

    if "bienes y servicios" in t:
        return "Bienes y servicios"

    return None


def encontrar_columnas_fecha(fila: list) -> list[dict]:
    out: list[dict] = []

    for idx, celda in enumerate(fila):
        fecha = convertir_fecha(celda)

        if fecha:
            out.append({"index": idx, "fecha": fecha})

    return out


def encontrar_concepto_en_fila(fila: list) -> str | None:
    for celda in fila:
        texto = limpiar_concepto(celda)

        if texto in CONCEPTOS_VALIDOS:
            return texto

    return None


def write_csv(out_file: Path, rows: list[dict]) -> None:
    out_file.parent.mkdir(parents=True, exist_ok=True)

    header = [
        "fecha",
        "region",
        "tipo_variacion",
        "bloque",
        "concepto",
        "valor",
        "archivo_origen",
        "fecha_descarga",
    ]

    with out_file.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=header)
        writer.writeheader()
        writer.writerows(rows)


def parse_sheet_to_rows(
    df: pd.DataFrame,
    archivo_origen: str,
) -> tuple[list[dict], str]:
    sheet_data = df.fillna("").values.tolist()
    tipo_variacion = detectar_tipo_variacion(sheet_data)

    region_actual = ""
    bloque_actual = ""
    columnas_fecha_actuales: list[dict] = []
    fecha_descarga = datetime.now().strftime("%Y-%m-%d")

    resultados: list[dict] = []

    for fila in sheet_data:
        columnas_fecha = encontrar_columnas_fecha(fila)

        if len(columnas_fecha) >= 3:
            columnas_fecha_actuales = columnas_fecha

        textos_fila = [
            normalizar_texto(x)
            for x in fila
            if normalizar_texto(x)
        ]

        for texto in textos_fila:
            if es_region(texto):
                region_actual = texto
                bloque_actual = ""

            bloque_detectado = detectar_bloque(texto)
            if bloque_detectado:
                bloque_actual = bloque_detectado

        concepto = encontrar_concepto_en_fila(fila)

        if (
            not concepto
            or not region_actual
            or not bloque_actual
            or not columnas_fecha_actuales
        ):
            continue

        for col in columnas_fecha_actuales:
            idx = col["index"]
            valor = convertir_numero(fila[idx] if idx < len(fila) else None)

            if valor is None:
                continue

            resultados.append(
                {
                    "fecha": col["fecha"],
                    "region": region_actual,
                    "tipo_variacion": tipo_variacion,
                    "bloque": bloque_actual,
                    "concepto": concepto,
                    "valor": valor,
                    "archivo_origen": archivo_origen,
                    "fecha_descarga": fecha_descarga,
                }
            )

    return resultados, tipo_variacion


def actualizar_ipc(anio: int, mes: int) -> None:
    archivo, content = buscar_y_descargar_ipc(anio, mes)

    DATA_DIR.mkdir(parents=True, exist_ok=True)
    (DATA_DIR / archivo).write_bytes(content)

    print("\nArchivo descargado correctamente y parseado en memoria:")
    print(archivo)

    xls = pd.ExcelFile(BytesIO(content))

    if not xls.sheet_names:
        raise RuntimeError("El archivo IPC no contiene hojas.")

    df0 = pd.read_excel(
        BytesIO(content),
        sheet_name=xls.sheet_names[0],
        header=None,
    )

    rows0, tipo0 = parse_sheet_to_rows(df0, archivo)
    write_csv(OUTPUT_FILE, rows0)

    print("Parse IPC terminado.")
    print(f"Hoja leída: {xls.sheet_names[0]}")
    print(f"Tipo de variación detectada: {tipo0}")
    print(f"Filas generadas: {len(rows0)}")
    print(f"Archivo generado: {OUTPUT_FILE}")

    if len(xls.sheet_names) > 1:
        df1 = pd.read_excel(
            BytesIO(content),
            sheet_name=xls.sheet_names[1],
            header=None,
        )

        rows1, tipo1 = parse_sheet_to_rows(df1, archivo)
        write_csv(OUTPUT_FILE_INTERANUAL, rows1)

        print("Parse IPC terminado.")
        print(f"Hoja leída: {xls.sheet_names[1]}")
        print(f"Tipo de variación detectada: {tipo1}")
        print(f"Filas generadas: {len(rows1)}")
        print(f"Archivo generado: {OUTPUT_FILE_INTERANUAL}")


def main() -> None:
    parser = argparse.ArgumentParser(description="Actualizar IPC INDEC")
    parser.add_argument("anio", type=int)
    parser.add_argument("mes", type=int)
    args = parser.parse_args()

    if args.mes < 1 or args.mes > 12:
        raise SystemExit("Mes inválido.")

    actualizar_ipc(args.anio, args.mes)


if __name__ == "__main__":
    try:
        main()
    except Exception as err:
        print("\nError actualizando IPC:")
        print(err)
        raise SystemExit(1) from err